File: packages/python/elizaos/bootstrap/actions/search_contacts.py
from __future__ import annotations

from dataclasses import dataclass, field
from typing import TYPE_CHECKING

from elizaos.bootstrap.utils.xml import parse_key_value_xml
from elizaos.types import (
    Action,
    ActionExample,
    ActionResult,
    Content,
    ModelType,
)

if TYPE_CHECKING:
    from elizaos.types import (
        HandlerCallback,
        HandlerOptions,
        IAgentRuntime,
        Memory,
        State,
    )


SEARCH_CONTACTS_TEMPLATE = """# Search Contacts

Current message: {{message}}
Sender: {{senderName}} (ID: {{senderId}})

Extract search criteria:
1. Categories (friend, family, colleague, acquaintance, vip, business)
2. Search terms (names or keywords)
3. Tags

<response>
<categories>comma-separated list of categories to filter by</categories>
<searchTerm>search term for names</searchTerm>
<tags>comma-separated list of tags</tags>
<intent>list, search, or count</intent>
</response>"""


@dataclass
class SearchContactsAction:
    name: str = "SEARCH_CONTACTS"
    similes: list[str] = field(
        default_factory=lambda: [
            "list contacts",
            "show contacts",
            "search contacts",
            "find contacts",
            "who are my friends",
        ]
    )
    description: str = "Search and list contacts in the rolodex"

    async def validate(
        self, runtime: IAgentRuntime, _message: Memory, _state: State | None = None
    ) -> bool:
        rolodex_service = runtime.get_service("rolodex")
        return rolodex_service is not None

    async def handler(
        self,
        runtime: IAgentRuntime,
        message: Memory,
        state: State | None = None,
        options: HandlerOptions | None = None,
        callback: HandlerCallback | None = None,
        responses: list[Memory] | None = None,
    ) -> ActionResult:
        from elizaos.bootstrap.services.rolodex import RolodexService

        rolodex_service = runtime.get_service("rolodex")
        if not rolodex_service or not isinstance(rolodex_service, RolodexService):
            return ActionResult(
                text="Rolodex service not available",
                success=False,
                values={"error": True},
                data={"error": "RolodexService not available"},
            )

        state = await runtime.compose_state(message, ["RECENT_MESSAGES"])

        prompt = runtime.compose_prompt_from_state(
            state=state,
            template=SEARCH_CONTACTS_TEMPLATE,
        )

        response = await runtime.use_model(ModelType.TEXT_SMALL, {"prompt": prompt})
        parsed = parse_key_value_xml(response)

        categories = None
        tags = None
        search_term = None

        if parsed:
            if parsed.get("categories"):
                categories = [c.strip() for c in str(parsed["categories"]).split(",") if c.strip()]
            if parsed.get("searchTerm"):
                search_term = str(parsed["searchTerm"])
            if parsed.get("tags"):
                tags = [t.strip() for t in str(parsed["tags"]).split(",") if t.strip()]

        contacts = await rolodex_service.search_contacts(
            categories=categories,
            tags=tags,
            search_term=search_term,
        )

        contact_details: list[dict[str, str]] = []
        for contact in contacts:
            entity = await runtime.get_entity(contact.entity_id)
            name = entity.name if entity and entity.name else "Unknown"
            contact_details.append(
                {
                    "id": str(contact.entity_id),
                    "name": name,
                    "categories": ",".join(contact.categories),
                    "tags": ",".join(contact.tags),
                }
            )

        if not contact_details:
            response_text = "No contacts found matching your criteria."
        else:
            response_text = f"I found {len(contact_details)} contact(s):\n"
            for detail in contact_details:
                response_text += f"- {detail['name']}"
                if detail["categories"]:
                    response_text += f" [{detail['categories']}]"
                response_text += "\n"

        if callback:
            await callback(Content(text=response_text.strip(), actions=["SEARCH_CONTACTS"]))

        return ActionResult(
            text=response_text.strip(),
            success=True,
            values={
                "count": len(contact_details),
            },
            data={
                "count": len(contact_details),
            },
        )

    @property
    def examples(self) -> list[list[ActionExample]]:
        return [
            [
                ActionExample(name="{{name1}}", content=Content(text="Show me all my friends")),
                ActionExample(
                    name="{{name2}}",
                    content=Content(
                        text="Here are your friends: Alice, Bob, Charlie",
                        actions=["SEARCH_CONTACTS"],
                    ),
                ),
            ],
        ]


search_contacts_action = Action(
    name=SearchContactsAction.name,
    similes=SearchContactsAction().similes,
    description=SearchContactsAction.description,
    validate=SearchContactsAction().validate,
    handler=SearchContactsAction().handler,
    examples=SearchContactsAction().examples,
)

File: packages/typescript/src/bootstrap/actions/searchContacts.ts
import { logger } from "../../logger.ts";
import type { RolodexService } from "../../services/rolodex.ts";
import type {
  Action,
  ActionResult,
  HandlerCallback,
  HandlerOptions,
  IAgentRuntime,
  Memory,
  State,
} from "../../types/index.ts";
import { ModelType } from "../../types/index.ts";
import { composePromptFromState, parseKeyValueXml } from "../../utils.ts";

interface SearchContactsXmlResult {
  categories?: string;
  searchTerm?: string;
  tags?: string;
  intent?: string;
}

const searchContactsTemplate = `# Search Contacts

Current message: {{message}}
Sender: {{senderName}} (ID: {{senderId}})

## Instructions
Extract the search criteria from the message:
1. Categories to filter by (friend, family, colleague, acquaintance, vip, business)
2. Search terms (names or keywords)
3. Tags to filter by
4. Any other filters mentioned

Do NOT include any thinking, reasoning, or <think> sections in your response.
Go directly to the XML response format without any preamble or explanation.

## Response Format
<response>
<categories>comma-separated list of categories to filter by</categories>
<searchTerm>search term for names</searchTerm>
<tags>comma-separated list of tags</tags>
<intent>list, search, or count</intent>
</response>

IMPORTANT: Your response must ONLY contain the <response></response> XML block above. Do not include any text, thinking, or reasoning before or after this XML block. Start your response immediately with <response> and end with </response>.`;

export const searchContactsAction: Action = {
  name: "SEARCH_CONTACTS",
  description: "Search and list contacts in the rolodex",
  similes: [
    "list contacts",
    "show contacts",
    "search contacts",
    "find contacts",
    "who are my friends",
    "list my colleagues",
    "show vip contacts",
    "find people named",
    "search for",
    "contacts list",
  ],
  examples: [
    [
      {
        name: "{{name1}}",
        content: { text: "Show me all my friends" },
      },
      {
        name: "{{name2}}",
        content: { text: "Here are your friends: Alice, Bob, Charlie" },
      },
    ],
    [
      {
        name: "{{name1}}",
        content: { text: "List my VIP contacts" },
      },
      {
        name: "{{name2}}",
        content: {
          text: "Your VIP contacts: Sarah (CEO), John (Key Client)",
        },
      },
    ],
    [
      {
        name: "{{name1}}",
        content: { text: "Find contacts named John" },
      },
      {
        name: "{{name2}}",
        content: {
          text: "I found 2 contacts named John: John Smith (colleague), John Doe (friend)",
        },
      },
    ],
  ],

  validate: async (
    runtime: IAgentRuntime,
    message: Memory,
    _state?: State,
  ): Promise<boolean> => {
    // Check if RolodexService is available
    const rolodexService = runtime.getService("rolodex") as RolodexService;
    if (!rolodexService) {
      logger.warn("[SearchContacts] RolodexService not available");
      return false;
    }

    // Check if message contains intent to search/list contacts
    const searchKeywords = [
      "list",
      "show",
      "search",
      "find",
      "contacts",
      "friends",
      "colleagues",
      "vip",
      "who",
    ];
    const messageText = message.content.text?.toLowerCase() || "";

    return searchKeywords.some((keyword) => messageText.includes(keyword));
  },

  handler: async (
    runtime: IAgentRuntime,
    message: Memory,
    state?: State,
    _options?: HandlerOptions,
    callback?: HandlerCallback,
  ): Promise<ActionResult | undefined> => {
    const rolodexService = runtime.getService("rolodex") as RolodexService;

    if (!rolodexService) {
      throw new Error("RolodexService not available");
    }

    // Build proper state for prompt composition
    if (!state) {
      state = {
        values: {},
        data: {},
        text: "",
      };
    }

    // Add our values to the state
    state.values = {
      ...state.values,
      message: message.content.text,
      senderId: message.entityId,
      senderName: state.values?.senderName || "User",
    };

    // Compose prompt to extract search criteria
    const prompt = composePromptFromState({
      state,
      template: searchContactsTemplate,
    });

    // Use LLM to extract search criteria
    const response = await runtime.useModel(ModelType.TEXT_SMALL, {
      prompt,
    });

    const parsedResponse = parseKeyValueXml<SearchContactsXmlResult>(response);

    // Build search criteria
    const criteria: {
      categories?: string[];
      tags?: string[];
      searchTerm?: string;
    } = {};

    if (parsedResponse?.categories) {
      criteria.categories = parsedResponse.categories
        .split(",")
        .map((c: string) => c.trim())
        .filter(Boolean);
    }

    if (parsedResponse?.searchTerm) {
      criteria.searchTerm = parsedResponse.searchTerm;
    }

    if (parsedResponse?.tags) {
      criteria.tags = parsedResponse.tags
        .split(",")
        .map((t: string) => t.trim())
        .filter(Boolean);
    }

    // Search contacts
    const contacts = await rolodexService.searchContacts(criteria);

    // Get entity names for each contact
    const contactDetails = await Promise.all(
      contacts.map(async (contact) => {
        const entity = await runtime.getEntityById(contact.entityId);
        return {
          contact,
          entity,
          name: entity?.names[0] || "Unknown",
        };
      }),
    );

    // Format response
    let responseText = "";

    if (contactDetails.length === 0) {
      responseText = "No contacts found matching your criteria.";
    } else if (parsedResponse?.intent === "count") {
      responseText = `I found ${contactDetails.length} contact${contactDetails.length !== 1 ? "s" : ""} matching your criteria.`;
    } else {
      // Group by category if searching all
      if (!criteria.categories || criteria.categories.length === 0) {
        const grouped = contactDetails.reduce(
          (acc, item) => {
            item.contact.categories.forEach((cat) => {
              if (!acc[cat]) acc[cat] = [];
              acc[cat].push(item);
            });
            return acc;
          },
          {} as Record<string, typeof contactDetails>,
        );

        responseText = `I found ${contactDetails.length} contact${contactDetails.length !== 1 ? "s" : ""}:\n\n`;

        for (const [category, items] of Object.entries(grouped)) {
          responseText += `**${category.charAt(0).toUpperCase() + category.slice(1)}s:**\n`;
          items.forEach((item) => {
            responseText += `- ${item.name}`;
            if (item.contact.tags.length > 0) {
              responseText += ` [${item.contact.tags.join(", ")}]`;
            }
            responseText += "\n";
          });
          responseText += "\n";
        }
      } else {
        // Simple list for specific category
        const categoryName = criteria.categories[0];
        responseText = `Your ${categoryName}s:\n`;
        contactDetails.forEach((item) => {
          responseText += `- ${item.name}`;
          if (item.contact.tags.length > 0) {
            responseText += ` [${item.contact.tags.join(", ")}]`;
          }
          responseText += "\n";
        });
      }
    }

    if (callback) {
      await callback({
        text: responseText,
        action: "SEARCH_CONTACTS",
        metadata: {
          count: contactDetails.length,
          criteria,
          success: true,
        },
      });
    }

    return {
      success: true,
      values: {
        count: contactDetails.length,
        criteria,
      },
      data: {
        count: contactDetails.length,
        criteria,
        contacts: contactDetails.map((d) => ({
          id: d.contact.entityId,
          name: d.name,
          categories: d.contact.categories,
          tags: d.contact.tags,
        })),
      },
      text: responseText,
    };
  },
};
